server/data/data-cassandra/src/main/java/org/apache/james/rrt/cassandra/CassandraRecipientRewriteTable.java

```diff
@@ -98,14 +101,26 @@ protected Mappings mapAddress(String user, Domain domain) {
         Preconditions.checkArgument(listSourcesSupportedType.contains(mapping.getType()),
             "Not supported mapping of type %s", mapping.getType());
 
-        SchemaVersion schemaVersion = cassandraSchemaVersionDAO.getCurrentSchemaVersion()
-            .block()
-            .orElse(CassandraSchemaVersionManager.MIN_VERSION);
-
-        if (schemaVersion.isBefore(MAPPINGS_SOURCES_SUPPORTED_VERSION)) {
+        if (isLegacy()) {
             return super.listSources(mapping);
         }
 
         return cassandraMappingsSourcesDAO.retrieveSources(mapping).toStream();
     }
+
+    private boolean isLegacy() {
```

Comment:
In my opinion, this kind of logic has nothing to do here. We must implement some logic in a MigrationSomething class that will handle caching logic.
Obviously, writing that kind of logic without writing any test should be warning sign that something is wrong.

Reply:
This changeset is already covered by tests on V6 & V7

I can write one more if you wish with a `CassandraRewriteTable` started on V6 and upgraded to V7.

BTW where would you put this logic?

Reply:
We may have the same problem in other classes.

What about having a `SchemaManager`, which can retrieve the version, with a one hour cache? (thanks to Reactor)